### PR TITLE
[SPARK-20412] Throw ParseException from visitNonOptionalPartitionSpec instead of returning null values.

### PR DESCRIPTION
**Pull request by juliuszsompolski**
## What changes were proposed in this pull request?

If a partitionSpec is supposed to not contain optional values, a ParseException should be thrown, and not nulls returned.
The nulls can later cause NullPointerExceptions in places not expecting them.

## How was this patch tested?

A query like "SHOW PARTITIONS tbl PARTITION(col1='val1', col2)" used to throw a NullPointerException.
Now it throws a ParseException.

### CODE REVIEW

**Comment by juliuszsompolski** [comment]
cc @cloud-fan

**Comment by hvanhovell** [comment]
ok to test

**Comment by hvanhovell** [comment]
Could you add a unit test to the DDLCommandSuite?

**Comment by cloud-fan** [comment]
looks like we need to fix a test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -214,8 +214,11 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
    * Create a partition specification map without optional values.
    */
   protected def visitNonOptionalPartitionSpec(
-      ctx: PartitionSpecContext): Map[String, String] = withOrigin(ctx) {
-    visitPartitionSpec(ctx).mapValues(_.orNull).map(identity)
+    ctx: PartitionSpecContext): Map[String, String] = withOrigin(ctx) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Nit: Indentation issues. Need to add extra two spaces

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -214,8 +214,11 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
...
+    visitPartitionSpec(ctx).map {
+      case (key, None) => throw new ParseException(s"Found empty key '$key'.", ctx)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: How about `Found an empty partition key '$key'`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLCommandSuite.scala [hunk 3]
```diff
@@ -558,12 +558,12 @@ class DDLCommandSuite extends PlanTest {
       tableIdent,
       Some("org.apache.class"),
       Some(Map("columns" -> "foo,bar", "field.delim" -> ",")),
-      Some(Map("test" -> null, "dt" -> "2008-08-08", "country" -> "us")))
+      Some(Map("test" -> "1", "dt" -> "2008-08-08", "country" -> "us")))
     val expected5 = AlterTableSerDePropertiesCommand(
       tableIdent,
       None,
       Some(Map("columns" -> "foo,bar", "field.delim" -> ",")),
-      Some(Map("test" -> null, "dt" -> "2008-08-08", "country" -> "us")))
```
Comment (juliuszsompolski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLCommandSuite.scala]: BTW: In AlterTableSerDePropertiesCommand that null would have caused a NullPointerExeception later, so not having this as null, but checked as ParseException is also a valid change here.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!

**Pull request closed without merging**